### PR TITLE
Fix `Map` representation in generated OpenAPI specs

### LINKED ISSUE #3048: [OpenAPI] `Map` fields are generating incorrect OpenAPI spec
This Endpoint definition
```scala
/**
 * This is the Input documentation
 */
final case class Input(headers: Map[String, String], body: String)

val testEndpoint = 
  (Endpoint(RoutePattern.POST / "test") ?? Doc.p("This is my 'POST /test' endpoint doc"))
    .in[Input]
    .out[Input](mediaType = MediaType.application.json, doc = Doc.p("this is the output doc"))

val spec: String =
  OpenAPIGen.fromEndpoints(
    title = "This is my OpenAPI doc title",
    version = "0.0.0",
    endpoints = List(testEndpoint)
  ).toJson

```
generates this OpenAPI spec for the `Input` type:
```json
      "Input": {
        "type": "object",
        "properties": {
          "body": {
            "type": "string"
          }
        },
        "additionalProperties": {
          "type": "string"
        },
        "required": [
          "headers",
          "body"
        ]
      },
```

The `headers` field is missing from the `properties` object and seems to be instead replaced with `additionalProperties` 🤔

@geeeezmo Do you also want to give this issue a try?

### PR DESCRIPTION
**Pull request by guizmaii**
Fixes #3048

### CODE REVIEW

File: zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala [hunk 1]
```diff
@@ -2812,6 +2812,13 @@ object OpenAPIGenSpec extends ZIOSpecDefault {
             |          "nestedOption" : {
             |            "$ref" : "#/components/schemas/Recursive"
             |          },
+            |          "nestedMap" : {
+            |            "type" : "object",
+            |            "properties" : {},
+            |            "additionalProperties" : {
+            |              "$ref" : "#/components/schemas/Recursive"
+            |            }
+            |          },
```
Comment (guizmaii) [inline zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:2821]: @hochgi @987Nabil I checked with the `openapi-generator-cli validate` cli and this definition is valid 🙂

With this fix, my TS code is correctly generated:
image

File: zio-http/shared/src/main/scala/zio/http/endpoint/openapi/JsonSchema.scala [hunk 2]
```diff
@@ -1212,10 +1212,30 @@ object JsonSchema {
...
+    /**
+     * Can't represent a case class and a Map at the same time
+     */
+    def isInvalid: Boolean = properties.nonEmpty && additionalProperties.isRight
+
     def addAll(value: Chunk[(java.lang.String, JsonSchema)]): Object =
       value.foldLeft(this) { case (obj, (name, schema)) =>
         schema match {
-          case Object(properties, additionalProperties, required) =>
+          case thatObj @ Object(properties, additionalProperties, required) if thatObj.isClosedDictionary =>
```
Comment (guizmaii) [inline zio-http/shared/src/main/scala/zio/http/endpoint/openapi/JsonSchema.scala:1238]: The issue was here. We were not treating the open dictionaries correctly 🙂

File: zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala (hunk 1 above)
Comment (hochgi) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:2821]: There's no need for `"properties" : {}`, its redundant. But no harm either I guess :)

Comment (guizmaii) [reply zio-http/jvm/src/test/scala/zio/http/endpoint/openapi/OpenAPIGenSpec.scala:2821]: Yes, it's not but I don't know how to remove it from the generated spec and the spec is valid with it, so yeah, no harm 🙂

**Comment by guizmaii** [comment]
Tests pass 🎉 
Ready for review 🚀

**Review by 987Nabil — APPROVED** [review]

**Pull request merged**